src/directive.ts

```diff
@@ -0,0 +1,267 @@
+
+  get options(): legacyLit.RenderOptions|undefined {
+    return undefined;
+  }
+  get name(): string {
+    return this.legacyPart.eventName;
+  }
+
+  get element(): Element {
+    return this.legacyPart.element;
+  }
+
+  /**
+   * If this attribute part represents an interpolation, this contains the
+   * static strings of the interpolation. For single-value, complete bindings,
+   * this is undefined.
+   */
+  get strings() {
+    return undefined;
+  }
+  get tagName() {
+    return this.element.tagName;
+  }
+
+  handleEvent(event: Event) {
+    this.legacyPart.handleEvent(event);
+  }
+}
+
+// no equivalent for ElementPart in v1
+
+function legacyPartToPart(part: legacyLit.Part): Part {
+  if (part instanceof legacyLit.NodePart) {
+    return new ChildPartImpl(part);
+  } else if (part instanceof legacyLit.EventPart) {
+    return new EventPartImpl(part);
+  } else if (part instanceof legacyLit.BooleanAttributePart) {
+    return new BooleanAttributePartImpl(part);
+  } else if (
+      part instanceof legacyLit.PropertyPart ||
+      part instanceof legacyLit.AttributePart) {
+    return new AttributePartImpl(part);
+  }
+  // ElementPartInfo doesn't exist in lit-html v1
+  throw new Error(`Unknown part type`);
+}
+
+/**
+ * Information about the part a directive is bound to.
+ *
+ * This is useful for checking that a directive is attached to a valid part,
+ * such as with directive that can only be used on attribute bindings.
+ */
+export type PartInfo = ChildPartInfo|AttributePartInfo;
+
+export interface DirectiveClass {
+  new(part: PartInfo): Directive;
+}
+
+/**
+ * This utility type extracts the signature of a directive class's render()
+ * method so we can use it for the type of the generated directive function.
+ */
+export type DirectiveParameters<C extends Directive> = Parameters<C['render']>;
+
+/**
+ * A generated directive function doesn't evaluate the directive, but just
+ * returns a DirectiveResult object that captures the arguments.
+ */
+export type DirectiveResult<C extends DirectiveClass = DirectiveClass> = {
+  /** @internal */
+  _$litDirective$: C;
+  /** @internal */
+  values: DirectiveParameters<InstanceType<C>>;
+};
+
+/**
+ * Base class for creating custom directives. Users should extend this class,
+ * implement `render` and/or `update`, and then pass their subclass to
+ * `directive`.
+ */
+export abstract class Directive {
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  constructor(_partInfo: PartInfo) {
+  }
+  abstract render(...props: Array<unknown>): unknown;
+  update(_part: Part, props: Array<unknown>): unknown {
+    return this.render(...props);
+  }
+}
+
+export function directive<C extends DirectiveClass>(directiveClass: C) {
+  const partToInstance =
+      new WeakMap<legacyLit.Part, readonly[Part, InstanceType<C>]>();
+  const result = legacyDirective((...props: unknown[]) => {
+    return (part: legacyLit.Part) => {
+      const cached = partToInstance.get(part);
```

Comment:
I think this is going to break in some nested directive situations. We need to try things like `guard([items], repeat(items, ...))`.

Reply:
Heya had a quick chat with peter and russell and we were a bit confused on what the concern is. Is it you think it'll write to the wrong location?

Additionally, we can't test with `guard` and `repeat` because they are not implemented using this class-based `Directive` impl. Additionally, assuming the user transferred their own directive to this and use it in one of our built-in ones, it should handle just fine.

The case where we think your concern about nested directives would come into play would be if a user implemented two of their own directives and nest them. I'd argue that's quite a fringe case.

I'd be happy to put a quick warning in a jsdoc saying this may not support certain advanced directive usage, but I'd like to understand the concern a bit more

Reply:
To clarify, I think:
* In 1.x, each type of directive hung its state off the part via weak map and it persisted even when directives changed and then changed back. Nested directives just worked by virtue of each having their own weak map on the part.
* In 2.x, each type of directive (by class) hangs its state off the part (or in the case of nesting, the nested directive), and its state is lost when directive type changes and then changes back.
* This PR currently matches the 2.x behavior as long as we don't consider nesting, but nesting will fail I think (but only for user directives, since the built-in directives don't use this, like Elliot said). Not sure if that's better or worse than caching by class type, which would make it more backward compatible (but doesn't get the clearing behavior of 2.x)

Reply:
It's not just nesting. This would fail:

```ts
html`${cond ? fooDirective() : barDirective()}`
```

because Foo and Bar directive classes would be mixed up.

Do something like, but with proper undefined handling:
```ts
const partToInstance = directiveCache.get(directiveClass);
const cached = partToInstance.get(part);
```

Reply:
Just checked. This does not fail because there is a weakmap per instance of `directive(...)` call rather than a persistent one. Added tests to show this